Alt text low code post (#228)
Quansight-Labs/quansight-labs-site: posts/2021/08/low-code-workshop.md

<!--

> .. title: Low-code contributions through github

Reviewer: Also `github` -> `GitHub`

Reviewer: in a few places

Author: i think i got em all

Believe it or not, we weren't lying to ourselves and others. Below we highlight our successes with running low-code sprints in `jupyter` and `numpy`.

### Jupyter accessibility group

There is a small group of the jupyter community focused on the improving the accessibility of jupyterlab and orbitting projects. 
We meet every other work to add small patches and fixes. currently, we're in the process of organizing workshops to advocate for accessibility and improve the quality of the entire jupyter experience for disabled scientists.

On an off week, between our normal syncs, we tested a low-code format for folks to participate in the jupyter project. 
During an hour, four people included in the project by suggesting changes to alt text for images in the `jupyterlab` documentation (https://github.com/isabela-pf/jupyterlab/pull/1), 
resulting in a collaborative commit to mainline with the help of @krassowski and @isabela-pf as champions for the shared work.(https://github.com/jupyterlab/jupyterlab/pull/10670)

We consider this event a success because we were able to remove `git` as a technical barrier to entry and support changes from multiple authors.

> ### `numpy` documentation new comers meeting

Reviewer 2: ```suggestion
### [NumPy documentation](https://numpy.org/doc/stable/) new comers meeting
```

> Our fledgling experience in one event made us *sort of* confident this could work for other projects. 

Reviewer 2: ```suggestion
Our fledgling experience in the Jupyter event made us *sort of* confident this could work for other projects. 
```

> So we contacted our `numpy` accessibility allies @melissawm and @marsbarlee to test this event at their numpy new comer's meeting.

Reviewer 2: ```suggestion
So we contacted our NumPy accessibility allies @melissawm and @marsbarlee to test this
hypothesis at the NumPy new comer's meeting, which [happens every other
Thursday](http://numpy-discussion.10968.n7.nabble.com/Newcomer-s-Hour-td49226.html).
```

> During the `numpy`, @marsbarlee provided a valuable [slideshow on how to write alt-text for scientific diagrams](https://docs.google.com/presentation/d/150vhbpGrtAc3ALhrS1a07lhEKCgevAY3ITh-4eCndDk/edit?usp=sharing). 

Reviewer 2: ```suggestion
During the meeting, @marsbarlee provided a valuable [slideshow on how to write alt-text for
scientific diagrams](https://docs.google.com/presentation/d/150vhbpGrtAc3ALhrS1a07lhEKCgevAY3ITh-4eCndDk/edit?usp=sharing). 
```

This involves looking at examples and looking at the image context.

> The progress from the `numpy` new comer's meeting can be seen in this issue https://github.com/melissawm/numpy/pull/27 .

Reviewer 2: ```suggestion
The progress from the NumPy new comer's meeting can be seen in this issue https://github.com/melissawm/numpy/pull/27 .
```

> A highlight, to us, is that we provided opportunities for both long time `numpy` fans and high school students a chance to see themselves in history of such a foundational tool like `numpy`. 

Reviewer 2: ```suggestion
A highlight, to us, is that we provided opportunities for both long time NumPy fans and high
school students a chance to see themselves in history of such a foundational tool like NumPy. 
```

<!--

> .. title: Low-code contributions through Github
> .. slug: low-code-contributions-through-Github

Reviewer 3: ```suggestion
.. title: Low-code contributions through GitHub
.. slug: low-code-contributions-through-GitHub
```

.. date: 2021-08-10 18:00:00 UTC-00:00

> .. author: Isabela Presedo-Floyd, Mars Lee, Melissa Weber Mendonça, Tony Fast

Reviewer 4: Is this listed alphabetical by first name? I know you don't tend to put your name first, but you really led writing this post.

Author: i don't know. this is an existential question. i originally forgot to put my name on it.

.. tags: Jupyter, NumPy, Accessibility
.. category:
.. link:
.. description:
.. type: text
-->

> # Low-code contributions through Github

Reviewer 3: ```suggestion
# Low-code contributions through GitHub
```

Healthy, inclusive communities are critical to impactful open source projects. 
A challenge for established projects is that the history and implicit technical debt increase the barrier to contribute to significant portions of code base.
The literacy of large code bases happens over time through incremental contributions, and we'll discuss a format that can begin this journey.

> At quansight labs, we are motivated to provide opportunities for new contributor's to experience open source community work regardless of their software literacy. 

Reviewer 3: Q,L Uppercase for Quansight and Labs maybe ?

Author: ```suggestion
At Quansight Labs, we are motivated to provide opportunities for new contributor's to experience open source community work regardless of their software literacy. 
```

> At Quansight Labs, we are motivated to provide opportunities for new contributor's to experience open source community work regardless of their software literacy. 

Reviewer 4: ```suggestion
At Quansight Labs, we are motivated to provide opportunities for new contributors to experience open source community work regardless of their software literacy. 
```

Community workshops are a common format for onboarding, but sometimes the outcome can be less than satisfactory for participants and organizers. 

> In these workshops, there are implicit challenges that need to be overcome to contribute to projects revision history like Git or setting up development environments. 

Reviewer 4: ```suggestion
In these workshops, there are implicit challenges that need to be overcome to contribute to projects' revision history like Git or setting up development environments. 
```

> Our goal with the following low-code workshop is to offer a way for folks to join a projects contributors list without the technical overhead. 

Reviewer 4: ```suggestion
Our goal with the following low-code workshop is to offer a way for folks to join a project's contributors list without the technical overhead. 
```

> To acheive this we'll discuss a format that relies solely on the Github web interface.

Reviewer 3: ```suggestion
To achieve this we'll discuss a format that relies solely on the GitHub web interface.
```

> 

Reviewer 3: ```suggestion

<!-- TEASER_END -->
```

So that not all the post is on the front page.

## Case study - collaboratively improving the accessibility of images in documentation

Our successful experiments are based on having new and seasoned contributors work together to add alternative text representations to images in project documentation. 
Each event lasts an hour with the following format:

1. pre-meeting preparation with a project contributor and meeting facilitator
2. a crash course in the specific topic - eg alt text for images
3. an introduction to a collaborative pull request created by a project contributor
4. group working session to suggest changes to specific files assigned in the pull request
5. a short review of the submissions by the core contributor
6. final review by the core contributor to submit a final pull request

> Accessibility is a focus for many members of quansight labs team so we've chosen this prompt to begin testing these workshops. 

Reviewer 5: ```suggestion
Accessibility is a focus for many members of Quansight Labs team so we've chosen this prompt to begin testing these workshops. 
```

> Accessibility is a focus for many members of Quansight Labs team so we've chosen this prompt to begin testing these workshops. 

Reviewer 4: ```suggestion
Accessibility is a focus for many members of Quansight Labs' team so we've chosen this prompt to begin testing these workshops. 
```

We'd love to hear about other prompts that may be good for low-code contributions.

### Pre-meeting preparation

Each event relies on a single pull request managed by a project contributor. 
We recommend using a fork to reduce noise to other contributors because we gonna be loud for an hour. 

Before the event the contributor makes a pull request identifying files that need changes related to the prompt, some projects may have rendered documentation that would also help guide the contributions. 

> The attendees will use these files to recommend changes to the code landing their aliases in the Github history. 

Reviewer 3: ```suggestion
The attendees will use these files to recommend changes to the code landing their aliases in the GitHub history. 
```

The facilitator will add an agenda for the event with learning materials that will prepare folks for the event.

### The crash course

The crash course is a short lightning talk style introduction to support the prompt folks will address during the hour. 
This short talk is important to prepare folks for the topic, to support contributors we want to remove the technical roadblocks that distract them from the community.

### Introduction to the pull request 

> To kick off the work, the core developer walks us through the pull request they've prepared and introduce to the different files. 

Reviewer 4: I'm not certain if you meant something else here, so my suggestion is a best guess at your thoughts.

```suggestion
To kick off the work, the core developer walks us through the pull request they've prepared and introduces different files. 
```

For reinforcement, the facilitator will share their screen to demonstrate the process of recommending changes to files in a pull request.

### Group work

Together the new and old contributors work together to make changes to relevant files to improve their quality. 

> These changes are made using Github's suggest changes feature that allows authors to make single line changes to code; a workflow perfect for alt text.

Reviewer 3: ```suggestion
These changes are made using GitHub's suggest changes feature that allows authors to make single line changes to code; a workflow perfect for alt text.
```

> ### group review

Reviewer 4: You capitalized the first letter of every other header.

```suggestion
### Group review
```

Over 40 minutes several contributors can make quite a few commits. 
The project contributor shares their screen to review the suggestions made over the course of the event. 
This portion becomes a critique to review the different contributions made.

### Bringing everything home

> All of the work we talked about happened on a contributors fork. 

Reviewer 4: ```suggestion
All of the work we talked about happened on a contributor's fork. 
```

> They now have a pull request will a gaggle of contributors. 

Reviewer 4: ```suggestion
They now have a pull request withl a gaggle of contributors. 
```

Their last job is to submit the pull request to the mainline project and see the work accepted into the code base.

## Success stories

Everyone loves an underdog story. 
What follows are there triumphs of some open source allies who thought they could run an hour sprint that allowed multiple folks to contribute to an open source project. 
Believe it or not, we weren't lying to ourselves and others. Below we highlight our successes with running low-code sprints in Jupyter and NumPy.

### Jupyter accessibility group

> There is a small group of the Jupyter community focused on the improving the accessibility of JupyterLab and orbitting projects. 

Reviewer 4: ```suggestion
There is a small group of the Jupyter community focused on the improving the accessibility of JupyterLab and orbiting projects. 
```

> We meet every other work to add small patches and fixes. currently, we're in the process of organizing workshops to advocate for accessibility and improve the quality of the entire Jupyter experience for disabled scientists.

Reviewer 4: ```suggestion
We meet every other work to add small patches and fixes. Currently, we're in the process of organizing workshops to advocate for accessibility and improve the quality of the entire Jupyter experience for disabled scientists.
```

On an off week, between our normal syncs, we tested a low-code format for folks to participate in the Jupyter project. 

> During an hour, four people included in the project by suggesting changes to alt text for images in the JupyterLab documentation (https://github.com/isabela-pf/jupyterlab/pull/1), 

Reviewer 4: ```suggestion
During an hour, four people were included in the project by suggesting changes to alt text for images in the JupyterLab documentation (https://github.com/isabela-pf/jupyterlab/pull/1), 
```

resulting in a collaborative commit to mainline with the help of @krassowski and @isabela-pf as champions for the shared work.(https://github.com/jupyterlab/jupyterlab/pull/10670)

We consider this event a success because we were able to remove `git` as a technical barrier to entry and support changes from multiple authors.

### [NumPy documentation](https://numpy.org/doc/stable/) new comers meeting

Our fledgling experience in the Jupyter event made us *sort of* confident this could work for other projects. 
So we contacted our NumPy accessibility allies @melissawm and @marsbarlee to test this

> hypothesis at the NumPy new comer's meeting, which [happens every other

Reviewer 4: Suggestion matches how NumPy advertises these meetings

```suggestion
hypothesis at the NumPy newcomer's meeting, which [happens every other
```

> hypothesis at the NumPy newcomer's meeting, which [happens every other

Reviewer 4: Adding mention of second event per request.

```suggestion
hypothesis at the NumPy newcomer's meetings, which [happen every other
```

Thursday](http://numpy-discussion.10968.n7.nabble.com/Newcomer-s-Hour-td49226.html).

> During the meeting, @marsbarlee provided a valuable [slideshow on how to write alt-text for

Reviewer 4: Alt text isn' hyphenated elsewhere in this post.

```suggestion
During the meeting, @marsbarlee provided a valuable [slideshow on how to write alt text for
```

> During the meeting, @marsbarlee provided a valuable [slideshow on how to write alt text for

Reviewer 4: Adding mention of second event per request.

```suggestion
During the first meeting, @marsbarlee provided a valuable [slideshow on how to write alt text for
```

scientific diagrams](https://docs.google.com/presentation/d/150vhbpGrtAc3ALhrS1a07lhEKCgevAY3ITh-4eCndDk/edit?usp=sharing). 

> This involves looking at examples and looking at the image context.

Reviewer 6: ```suggestion
Writing alt-text for scientific diagrams in the documentation poses a unique challenge, as conventional advice for writing alt-text is not sufficient. Abstract concepts such as mathematics can be difficult to describe visually.

 We tackled this in the workshop by comparing different examples, such as the similarities and differences in writing alt-text for a Venn Diagram versus a bar chart. We also took cue from the surrounding context of the diagram, such as the code comments, to find the most relevant information to visually describe.
```

> The progress from the NumPy new comer's meeting can be seen in this issue https://github.com/melissawm/numpy/pull/27 .

Reviewer 4: ```suggestion
The progress from the NumPy newcomer's meeting can be seen in this issue https://github.com/melissawm/numpy/pull/27 .
```

> 

Reviewer 4: ```suggestion
@melissawm led the second iteration of this event to complete more alt text in the main NumPy documentation as well as beginning to cover [NumPy tutorials](https://numpy.org/numpy-tutorials/) with a new group of people. Including NumPy tutorials, which are originally written as Jupyter notebooks, also brought in questions about handling image or other types of media descriptions for cell outputs the tutorial produced. Working across repositories and markup languages, the community gathered to create another chunk of progress in more accessible documentation across the project. 
```

Author: @MarsBarLee @melissawm i was wondering it y'all might want to add anything here. @isabela-pf thanks for this update

Reviewer 6: ```suggestion
Melissa led the second iteration of this event to complete more alt text in the main NumPy documentation as well as beginning to cover [NumPy tutorials](https://numpy.org/numpy-tutorials/) with a new group of people. By hosting a workshop with another project, we saw the similar yet different challenges NumPy faced compared to Jupyter,

By working on the NumPy tutorials, which are originally written as Jupyter notebooks, we faced new questions, such as how to handling rendered live charts, which are not in still images formats such as JPEG. We faced other types of media, such as cell input and outputs.

We faced a variety of markup languages, including reST, Markdown and HTML, each with it's own quirks and slight variation in implementing alt-text.

Working across repositories and markup languages, the community gathered once again to create more accessible documentation across the open-source space, one contribution at a time.
```

Reviewer 6: I just added on to Isabela's suggestion, with some more detail. I'm also not sure if you want to keep @github-handle vs using person-name. I went with person-name here but you can change it.

A highlight, to us, is that we provided opportunities for both long time NumPy fans and high

> school students a chance to see themselves in history of such a foundational tool like NumPy. 

Reviewer 4: ```suggestion
school students to see themselves in history of such a foundational tool like NumPy. 
```

Moreover, attendees were still adding alt text contributions after the event concluded.

## Conclusion

> Using Github's suggested change system is a way to support low-code, single-line contributions to different open source code bases. 

Reviewer 3: ```suggestion
Using GitHub's suggested change system is a way to support low-code, single-line contributions to different open source code bases. 
```

The event's collaborative hands-on format offers time for the community to interact in a meaningful way and see the fruits of their labor immediately.

> Please reach out to us if you or your project are interested in running event like this. 

Reviewer 4: We might want to say and/or link to how to reach out. I don't know what the preferred way is, though. Is this direct to a community member listed above, or to Quansight Labs, or something else?

Author: maybe we can say to respond in the comments?

Reviewer 4: I don't think we have commenting on the blog? Or am I not seeing it?

Reviewer 3: There is commenting on the blog now. It does requires a github account. 
See https://github.com/Quansight-Labs/quansight-labs-site/pull/223 for background. 

Comments are actually reply backed by a GitHub issues – which is created once the first comment is posted.

Author: thanks for landing #233 that is rad!

<!--
.. title: Low-code contributions through GitHub
.. slug: low-code-contributions-through-GitHub

> .. date: 2021-08-10 18:00:00 UTC-00:00

Author: ```suggestion
.. date: 2021-09-8 18:00:00 UTC-00:00
```
